### PR TITLE
Fixing DB not initialized when PN arrives

### PR DESCRIPTION
**Pull request by milos1290**
What              | Where/Who
------------------|----------------------------------------
JIRA Issue        | [JIRA_TICKET_ID](https://leanplum.atlassian.net/browse/E2-1970)
People Involved   | @milos1290, @hrishileanplum 

[Notes on PR descriptions](https://leanplum.atlassian.net/wiki/spaces/PR/pages/288424408/Creating+a+GitHub+Pull+Request)

## Background
This PR fixes 2 issues
1) when PN arrives and we need to pull new content, DB is not initialised because initialisation happens in `start` and app is not running.
2) when PN arrives we are offloading a network request from IntentService to an async task, service will be killed and kill our process.

## Implementation

1) Making DB manager singleton
2) conditionally decide whether to run task async or on current thread.

## Testing steps

* Install the app
* Kill the app from task manager
* Create IAM
* Send push that chains to IAM from previous step

## Is this change backwards-compatible?

Yes

### CODE REVIEW

File: AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java [hunk 1]
```diff
@@ -48,53 +48,58 @@
...
   private static final String COLUMN_DATA = "data";
   private static final String KEY_ROWID = "rowid";
 
-  private static SQLiteDatabase database;
-  private static LeanplumDataBaseManager databaseManager;
-  private static ContentValues contentValues = new ContentValues();
+  private SQLiteDatabase database;
+  private LeanplumDataBaseManager databaseManager;
+  private ContentValues contentValues = new ContentValues();
+  private boolean sendErrorLogs = false;
```
Comment (hrishileanplum) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java]: Any reason for "sendErrorLogs" rename?

File: AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java [hunk 2]
```diff
@@ -48,53 +48,58 @@
...
+    }
+    return instance;
+  }
+
   /**
    * Inserts event to event table.
    *
    * @param event String with json of event.
    */
-  static void insertEvent(String event) {
```
Comment (hrishileanplum) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java]: To confirm here. So since LeanplumEventDataManager is singleton, you dont need static?

File: AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java [hunk 3]
```diff
@@ -158,33 +163,41 @@ static void deleteEvents(int count) {
...
     } catch (Throwable t) {
       handleSQLiteError("Unable to get a number of rows in the table.", t);
     }
     return count;
   }
 
+  /**
+   * Whether we are going to send error log or not.
+   */
+  boolean willSendErrorLogs() {
```
Comment (hrishileanplum) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java]: Why is this needed since its default value? Where do we change the value .

File: AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java [hunk 4]
```diff
@@ -239,6 +252,8 @@ private static void migrateFromSharedPreferences(SQLiteDatabase db) {
 
         editor.remove(Constants.Defaults.COUNT_KEY);
 
+        ContentValues contentValues = new ContentValues();
```
Comment (hrishileanplum) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java]: Can you please explain the need of ContentValues here?

File: AndroidSDKCore/src/main/java/com/leanplum/internal/RequestOld.java [hunk 5]
```diff
@@ -516,20 +520,29 @@ private void sendNow() {
...
+          } catch (Throwable t) {
+            Util.handleException(t);
+          }
+          return null;
         }
-        return null;
+      });
+    } else {
+      try {
+        sendRequests();
```
Comment (hrishileanplum) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/RequestOld.java]: Both if and else do the same thing?           try {
            sendRequests();
          } catch (Throwable t) {
            Util.handleException(t);
          }

File: AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java [hunk 6]
```diff
@@ -625,7 +625,7 @@ private static void checkAndStartNotificationsModules() {
 
   private static void startHelper(
       String userId, final Map<String, ?> attributes, final boolean isBackground) {
-    LeanplumEventDataManager.init(context);
+    LeanplumEventDataManager.sharedInstance();
```
Comment (hrishileanplum) [inline AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java]: Can you please explain why initialization wasnt happening? I see init here correct

File: AndroidSDKTests/src/test/java/com/leanplum/internal/RequestOldTest.java [hunk 7]
```diff
@@ -262,7 +262,7 @@ public void testRemoveIrrelevantBackgroundStartRequests() throws NoSuchMethodExc
   @Test
   public void testJsonEncodeUnsentRequestsWithExceptionLargeNumbers() throws NoSuchMethodException,
```
Comment (hrishileanplum) [inline AndroidSDKTests/src/test/java/com/leanplum/internal/RequestOldTest.java]: You need extra test cases, specific to initialization.

File: AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java (hunk 1 above)
Comment (milos1290) [reply AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java]: Just to make it consistent, nothing else.

File: AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java (hunk 2 above)
Comment (milos1290) [reply AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java]: Yeah, its singleton now and no need for static methods.

File: AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java (hunk 3 above)
Comment (milos1290) [reply AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java]: We use the flag in case we fail to save data into SQL so we could send error logs from database

File: AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java (hunk 4 above)
Comment (milos1290) [reply AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumEventDataManager.java]: It used static one from outer class, since its a private variable now its not accessible anymore.

File: AndroidSDKCore/src/main/java/com/leanplum/internal/RequestOld.java (hunk 5 above)
Comment (milos1290) [reply AndroidSDKCore/src/main/java/com/leanplum/internal/RequestOld.java]: Not really, the first one is offloading request into background task, the else part is running it on current thread (which is background thread)

File: AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java (hunk 6 above)
Comment (milos1290) [reply AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java]: As you can see, initialization happens in `startHelper` which is actually causing the issues when app is not running (i.e push notification received).

File: AndroidSDKCore/src/main/java/com/leanplum/internal/RequestOld.java (hunk 5 above)
Comment (hrishileanplum) [reply AndroidSDKCore/src/main/java/com/leanplum/internal/RequestOld.java]: Can you please let me know why is that point above? Why couldnt we just run in background thread, no matter what the parent thread is?

Comment (milos1290) [reply AndroidSDKCore/src/main/java/com/leanplum/internal/RequestOld.java]: Yeah it doesn't really matter, we can always create new async task, but in that case I'll need to synchronise call to forceContentUpdate in LeanplumPushService, to make sure that IntentService doesn't get killed.

Comment (hrishileanplum) [reply AndroidSDKCore/src/main/java/com/leanplum/internal/RequestOld.java]: I would prefer that approach please. Its way more cleaner than handling these If else conditions. Can you please make that change?
